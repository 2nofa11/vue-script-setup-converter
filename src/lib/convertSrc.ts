import {
  CallExpression,
  ScriptTarget,
  SyntaxKind,
  Project,
  Node,
  ts,
} from "ts-morph";
import { parse } from "@vue/compiler-sfc";
import { getNodeByKind } from "./helpers/node";
import { hasNamedImportIdentifier } from "./helpers/module";
import { convertImportDeclaration } from "./converter/importDeclarationConverter";
import { convertPageMeta } from "./converter/pageMetaConverter";
import { convertProps } from "./converter/propsConverter";
import { convertSetup } from "./converter/setupConverter";
import { convertEmits } from "./converter/emitsConverter";
import { convertComponents } from "./converter/componentsConverter";
import { convertDefineOptions } from "./converter/defineOptionsConverter";
import { genImport } from "knitwork";

export const isAlreadyScriptSetup = (input: string): boolean => {
  const { descriptor } = parse(input);
  return !!descriptor.scriptSetup;
};

export const hasScript = (input: string): boolean => {
  const { descriptor } = parse(input);
  return !!descriptor.script;
};

export const convertSrc = (input: string) => {
  if (isAlreadyScriptSetup(input)) {
    return input;
  }

  if (!hasScript(input)) {
    return input;
  }

  const {
    descriptor: { script },
  } = parse(input);

  const project = new Project({
    compilerOptions: {
      target: ScriptTarget.Latest,
    },
    useInMemoryFileSystem: true,
  });

  const sourceFile = project.createSourceFile("s.tsx", script?.content ?? "");
  const lang = script?.lang ?? "js";

  const callexpression = getNodeByKind(sourceFile, SyntaxKind.CallExpression);

  if (!callexpression) {
    throw new Error("defineComponent is not found.");
  }
  if (!Node.isCallExpression(callexpression)) {
    throw new Error("defineComponent is not found.");
  }

  if (!isDefineComponent(callexpression)) {
    throw new Error("defineComponent is not found.");
  }

  const importMap = convertImportDeclaration(sourceFile) ?? "";
  const pageMeta = convertPageMeta(callexpression, lang) ?? "";
  const propsResult = convertProps(callexpression, lang);
  const props = propsResult.code ?? "";
  const emits = convertEmits(callexpression, lang) ?? "";
  const statement = convertSetup(callexpression, propsResult.propNames) ?? "";
  const components = convertComponents(callexpression) ?? "";
  const defineOptions = convertDefineOptions(callexpression) ?? "";

  const hasDynamicImport = components.includes("defineAsyncComponent");
  const hasUseAttrs = statement.includes("useAttrs()");
  const hasUseSlots = statement.includes("useSlots()");

  const statements = project.createSourceFile("new.tsx");

  // Add necessary imports
  if (importMap.length > 0) {
    if (
      hasDynamicImport &&
      !importMap[0].importSpecifiers.includes("defineAsyncComponent")
    ) {
      importMap[0].importSpecifiers.push("defineAsyncComponent");
    }

    if (hasUseAttrs && !importMap[0].importSpecifiers.includes("useAttrs")) {
      importMap[0].importSpecifiers.push("useAttrs");
    }

    if (hasUseSlots && !importMap[0].importSpecifiers.includes("useSlots")) {
      importMap[0].importSpecifiers.push("useSlots");
    }
  }

  // Only add import statements if there are valid import specifiers
  const validImports = importMap.filter((x) => x.importSpecifiers.length > 0);
  if (validImports.length > 0) {
    statements.addStatements(
      validImports.map((x) => genImport(x.moduleSpecifier, x.importSpecifiers))
    );
  }

  statements.addStatements(
    sourceFile
      .getStatements()
      .filter((state) => {
        if (Node.isExportAssignment(state)) return false;
        if (Node.isImportDeclaration(state)) {
          const moduleSpecifier = state.getModuleSpecifierValue();
          // Imports from vue and #imports are already processed by convertImportDeclaration, so exclude them
          // However, preserve type-only imports
          if (
            ["vue", "#imports"].includes(moduleSpecifier) &&
            !state.isTypeOnly()
          )
            return false;
          if (
            hasNamedImportIdentifier(state, "defineComponent") ||
            hasNamedImportIdentifier(state, "defineNuxtComponent")
          )
            return false;
        }

        return true;
      })
      .map((x) => {
        return x.getText();
      })
  );

  statements.addStatements(components);

  if (defineOptions) {
    statements.addStatements(defineOptions);
  }

  if (isDefineNuxtComponent(callexpression)) {
    statements.addStatements(pageMeta);
  }

  statements.addStatements(props);
  statements.addStatements(emits);
  statements.addStatements(statement);

  statements.formatText({
    semicolons: ts.SemicolonPreference.Insert,
    indentSize: 2,
  });

  const scriptSetupTag =
    lang === "ts" ? '<script setup lang="ts">' : "<script setup>";

  // Get full text and ensure proper line breaks between statements
  let convertedCode = statements.getFullText().trim();

  // Fix the specific issue where type Props and const are on the same line
  convertedCode = convertedCode.replace(/};\s*const\s+/g, "};\nconst ");

  // Handle empty component case - ensure at least one newline for minimal valid script setup
  if (convertedCode === "") {
    return `${scriptSetupTag}\n</script>`;
  }

  return `${scriptSetupTag}\n${convertedCode}\n</script>`;
};

const isDefineComponent = (node: CallExpression) => {
  if (!Node.isIdentifier(node.getExpression())) {
    return false;
  }

  return (
    node.getExpression().getText() === "defineComponent" ||
    node.getExpression().getText() === "defineNuxtComponent"
  );
};

const isDefineNuxtComponent = (node: CallExpression) => {
  if (!Node.isIdentifier(node.getExpression())) {
    return false;
  }
  return node.getExpression().getText() === "defineNuxtComponent";
};
