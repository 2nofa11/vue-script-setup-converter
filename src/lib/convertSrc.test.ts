import { expect, describe, it } from "vitest";
import { convertSrc, isAlreadyScriptSetup, hasScript } from "./convertSrc";

describe("snapshot", () => {
  it("lang=js", () => {
    const output = convertSrc(`<script>
import { defineComponent, toRefs, computed, ref } from 'vue';

export default defineComponent({
  name: 'HelloWorld',
  emits: {
    change: null
  },
  props: {
    msg: {
      type: String,
      default: 'HelloWorld'
    },
    foo: {
      type: String,
      required: true,
    }
  },
  setup(props, ctx) {
    const { msg, foo } = toRefs(props);
    const newMsg = computed(() => msg.value + '- HelloWorld');

    const count = ref(0);

    return {
      count,
      newMsg
    }
  }
})
</script>`);
    expect(output).toMatchSnapshot();
  });

  it("lang=ts", () => {
    const output = convertSrc(`<script lang="ts">
import { defineComponent, toRefs, computed, ref } from 'vue';

export default defineComponent({
  name: 'HelloWorld',
  emits: {
    change: (value: number) => true,
  },
  props: {
    msg: {
      type: String,
      default: 'HelloWorld'
    },
    foo: {
      type: String
      required: true,
    }
  },
  setup(props, ctx) {
    const { msg, foo } = toRefs(props);
    const newMsg = computed(() => msg.value + '- HelloWorld');

    const count = ref(0);
    ctx.emit("change", 124);

    return {
      count,
      newMsg
    }
  }
})
</script>`);
    expect(output).toMatchSnapshot();
  });

  it("defineNuxtComponent", () => {
    const output = convertSrc(`<script lang="ts">
import { defineNuxtComponent, useNuxtApp } from '#imports';

export default defineNuxtComponent({
  name: 'HelloWorld',
  layout: 'test-layout',
  middleware: 'test-middleware',
  setup(props, ctx) {
    const { $client } = useNuxtApp();

    const onSubmit = () => {
      console.log('onSubmit')
    }

    return {
      onSubmit,
    }
  }
})
</script>
`);
    expect(output).toMatchSnapshot();
  });

  it("empty setup context", () => {
    const output = convertSrc(`<script lang="ts">
import { defineComponent, toRefs, computed } from 'vue';

export default defineComponent({
  name: 'HelloWorld',
  props: {
    msg: {
      type: String,
      default: 'HelloWorld'
    }
  },
  setup(props) {
    const { msg } = toRefs(props);
    const newMsg = computed(() => msg.value + '- HelloWorld');

    return {
      newMsg
    }
  }
})
</script>`);
    expect(output).toMatchInlineSnapshot(
      `
      "<script setup lang="ts">
      import { toRefs, computed } from "vue";
      type Props = { msg?: string; };
      const { msg = 'HelloWorld' } = defineProps<Props>();

      const { msg } = toRefs(props);
      const newMsg = computed(() => msg.value + '- HelloWorld');
      </script>"
    `
    );
  });

  it("props no toRefs", () => {
    const output = convertSrc(`<script lang="ts">
    import type { PropType } from 'vue';
import { defineComponent, computed } from 'vue';

export default defineComponent({
  name: 'HelloWorld',
  props: {
    msg: {
      type: String,
      default: 'HelloWorld'
    }
  },
  setup(props) {
    const newMsg = computed(() => props.msg + '- HelloWorld');

    return {
      newMsg
    }
  }
})
</script>`);
    expect(output).toMatchInlineSnapshot(
      `
      "<script setup lang="ts">
      import { computed } from "vue";
      import type { PropType } from 'vue';
      type Props = { msg?: string; };
      const { msg = 'HelloWorld' } = defineProps<Props>();

      const newMsg = computed(() => msg + '- HelloWorld');
      </script>"
    `
    );
  });

  it("should be converted to defineAsyncComponent", () => {
    const output = convertSrc(`<script>
import { defineComponent } from 'vue';
import HelloWorld from './HelloWorld.vue';

export default defineComponent({
  components: {
    HelloWorld,
    MyComp: () => import('./MyComp.vue'),
    Foo: () => import('./Foo.vue'),
  }
  })
  </script>`);
    expect(output).toMatchInlineSnapshot(`
      "<script setup>
      import { defineAsyncComponent } from "vue";
      import HelloWorld from './HelloWorld.vue';
      const MyComp = defineAsyncComponent(() => import('./MyComp.vue'));
      const Foo = defineAsyncComponent(() => import('./Foo.vue'));
      </script>"
    `);
  });

  it("should not be empty import", () => {
    const output = convertSrc(`<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  emits: ['click'],
  setup(_, { emit }) {
    const click = () => {
      emit('click')
    }
    return { click }
  },
})
</script>`);
    expect(output).toMatchInlineSnapshot(`
      "<script setup lang="ts">
      const emit = defineEmits(['click']);

      const click = () => {
        emit('click');
      };
      </script>"
    `);
  });

  it("should convert useAttrs and inheritAttrs", () => {
    const output = convertSrc(`<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  inheritAttrs: false,
  setup(_, { attrs }) {
    return {
      attrs,
    }
  },
})
</script>`);
    expect(output).toMatchInlineSnapshot(`
      "<script setup lang="ts">
      import { useAttrs } from "vue";
      defineOptions({
        inheritAttrs: false
      });
      const attrs = useAttrs();
      </script>"
    `);
  });

  it("should replace props references with destructured names", () => {
    const output = convertSrc(`<script lang="ts">
import { computed, defineComponent } from 'vue'

export default defineComponent({
  props: {
    from: {
      type: Number,
      required: true,
    },
    to: {
      type: Number,
      required: true,
    },
    total: {
      type: Number,
      required: true,
    },
    isLoading: {
      type: Boolean,
      default: false,
    },
  },
  setup(props) {
    const isFirst = computed(() => props.from === 1)
    const isLast = computed(() => props.to === props.total)
    const isDisabled = computed(() => props.isLoading)

    return {
      isFirst,
      isLast,
      isDisabled,
    }
  },
})
</script>`);
    expect(output).toContain("const isFirst = computed(() => from === 1)");
    expect(output).toContain("const isLast = computed(() => to === total)");
    expect(output).toContain("const isDisabled = computed(() => isLoading)");
    expect(output).toContain(
      "const { from, to, total, isLoading = false } = defineProps<Props>();"
    );
  });

  it("should convert empty defineComponent to minimal script setup", () => {
    const output = convertSrc(`<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({})
</script>`);
    expect(output).toMatchInlineSnapshot(`
      "<script setup lang="ts">
      </script>"
    `);
  });

  it("should convert empty defineComponent without lang attribute", () => {
    const output = convertSrc(`<script>
import { defineComponent } from 'vue'

export default defineComponent({})
</script>`);
    expect(output).toMatchInlineSnapshot(`
      "<script setup>
      </script>"
    `);
  });

  describe("import duplication issues", () => {
    it("should not duplicate imports from #imports", () => {
      const input = `<template>
  <div class="searchNoResult" >
    <div class="searchNoResult-empty">
      <div>
        <UserIcon />
        <div>
          No results found.<br />Please try different search criteria.
        </div>
        <Button
          @click="goSearch"
          >Edit Criteria</Button
        >
      </div>
    </div>
    <div class="globalFooter">
      <PortalTarget name="global-footer" />
    </div>
  </div>
</template>

<script lang="ts">
import { useRoute, useRouter } from '#imports'
import { defineComponent } from 'vue'

import UserIcon from '~/assets/icons/user-icon.svg'

export default defineComponent({
  name: 'SearchNoResult',
  components: {
    UserIcon,
  },
  setup(_) {
    const router = useRouter()
    const route = useRoute()

    const goSearch = () => {
      const query = { ...route.query }
      delete query.page

      router.push({
        path: '/search',
        query,
      })
    }

    return {
      goSearch,
    }
  },
})
</script>`;

      const result = convertSrc(input);

      // Check that useRoute and useRouter are imported only once
      const importLines = result
        .split("\n")
        .filter((line) => line.includes("import"));
      const importsFromSharpImports = importLines.filter(
        (line) => line.includes("#imports") || line.includes('"#imports"')
      );

      // Should have only one import from #imports
      expect(importsFromSharpImports).toHaveLength(1);

      // Check that the import contains both useRoute and useRouter
      const importLine = importsFromSharpImports[0];
      expect(importLine).toContain("useRoute");
      expect(importLine).toContain("useRouter");

      // Check that there are no duplicate import lines
      const uniqueImportLines = [...new Set(importLines)];
      expect(importLines).toHaveLength(uniqueImportLines.length);
    });

    it("should not duplicate imports from vue", () => {
      const input = `<script lang="ts">
import { defineComponent, ref, computed } from 'vue'
import { useRoute } from '#imports'

export default defineComponent({
  setup() {
    const count = ref(0)
    const doubled = computed(() => count.value * 2)
    
    return {
      count,
      doubled
    }
  }
})
</script>`;

      const result = convertSrc(input);

      // Check that vue imports are not duplicated
      const importLines = result
        .split("\n")
        .filter((line) => line.includes("import"));
      const vueImports = importLines.filter(
        (line) => line.includes('from "vue"') || line.includes("from 'vue'")
      );

      // Should have only one import from vue (if any)
      expect(vueImports.length).toBeLessThanOrEqual(1);

      // Check for duplicate lines
      const uniqueImportLines = [...new Set(importLines)];
      expect(importLines).toHaveLength(uniqueImportLines.length);
    });

    it("should preserve other imports correctly", () => {
      const input = `<script lang="ts">
import { useRoute, useRouter } from '#imports'
import { defineComponent } from 'vue'
import UserIcon from '~/assets/icons/user-icon.svg'
import { someUtility } from '~/utils/helper'

export default defineComponent({
  components: {
    UserIcon,
  },
  setup() {
    const router = useRouter()
    const route = useRoute()
    
    return {
      router,
      route
    }
  }
})
</script>`;

      const result = convertSrc(input);

      // Should preserve non-vue/non-#imports imports
      expect(result).toContain("from '~/assets/icons/user-icon.svg'");
      expect(result).toContain("from '~/utils/helper'");

      // Should not duplicate imports
      const importLines = result
        .split("\n")
        .filter((line) => line.includes("import"));
      const uniqueImportLines = [...new Set(importLines)];
      expect(importLines).toHaveLength(uniqueImportLines.length);
    });
  });
});

describe("script setup detection and skip", () => {
  it("should detect script setup files", () => {
    const scriptSetupFile = `<template>
  <div>{{ msg }}</div>
</template>

<script setup lang="ts">
const msg = 'Hello World'
</script>`;

    expect(isAlreadyScriptSetup(scriptSetupFile)).toBe(true);
  });

  it("should not detect regular script as script setup", () => {
    const regularScriptFile = `<template>
  <div>{{ msg }}</div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  setup() {
    const msg = 'Hello World'
    return { msg }
  }
})
</script>`;

    expect(isAlreadyScriptSetup(regularScriptFile)).toBe(false);
  });

  it("should skip conversion for script setup files", () => {
    const scriptSetupFile = `<template>
  <div>{{ msg }}</div>
</template>

<script setup lang="ts">
const msg = 'Hello World'
</script>`;

    const result = convertSrc(scriptSetupFile);
    expect(result).toBe(scriptSetupFile);
  });
});

describe("no script section handling", () => {
  it("should detect files with script section", () => {
    const fileWithScript = `<template>
  <div>Hello World</div>
</template>

<script>
export default {
  name: 'HelloWorld'
}
</script>`;

    expect(hasScript(fileWithScript)).toBe(true);
  });

  it("should detect files without script section", () => {
    const fileWithoutScript = `<template>
  <div>Hello World</div>
</template>

<style scoped>
div {
  color: red;
}
</style>`;

    expect(hasScript(fileWithoutScript)).toBe(false);
  });

  it("should skip conversion for files without script section", () => {
    const templateOnlyFile = `<template>
  <div>Hello World</div>
</template>

<style scoped>
div {
  color: red;
}
</style>`;

    const result = convertSrc(templateOnlyFile);
    expect(result).toBe(templateOnlyFile);
  });

  it("should skip conversion for template-only files", () => {
    const templateOnlyFile = `<template>
  <div class="container">
    <h1>Welcome</h1>
    <p>This is a template-only component</p>
  </div>
</template>`;

    const result = convertSrc(templateOnlyFile);
    expect(result).toBe(templateOnlyFile);
  });
});
