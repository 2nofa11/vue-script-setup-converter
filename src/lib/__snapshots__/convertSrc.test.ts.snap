// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot > defineNuxtComponent 1`] = `
"<script setup lang="ts">
import { useNuxtApp } from "#imports";
definePageMeta({
  name: 'HelloWorld',
  layout: 'test-layout',
  middleware: 'test-middleware'
});

const { $client } = useNuxtApp();

const onSubmit = () => {
  console.log('onSubmit');
};
</script>"
`;

exports[`snapshot > lang=js 1`] = `
"<script setup>
import { toRefs, computed, ref } from "vue";
const props = defineProps({
  msg: {
    type: String,
    default: 'HelloWorld'
  }, foo: {
    type: String,
    required: true,
  }
});
const emit = defineEmits({ change: null });

const { msg, foo } = toRefs(props);
const newMsg = computed(() => msg.value + '- HelloWorld');

const count = ref(0);
</script>"
`;

exports[`snapshot > lang=ts 1`] = `
"<script setup lang="ts">
import { toRefs, computed, ref } from "vue";
type Props = {
  msg?: string;
  foo: string;
};
const { msg = 'HelloWorld', foo } = defineProps<Props>();
const emit = defineEmits<{ (e: 'change', value: number): void; }>();

const { msg, foo } = toRefs(props);
const newMsg = computed(() => msg.value + '- HelloWorld');

const count = ref(0);
emit("change", 124);
</script>"
`;
